src/lib/cpd/CpdEngine.ts

```diff
@@ -0,0 +1,244 @@
+import { uxEvents } from '../ScannerEvents';
+
+
+export class CpdEngine extends AbstractRuleEngine {
+
+	public static ENGINE_ENUM: ENGINE = ENGINE.CPD;
+	public static ENGINE_NAME: string = ENGINE.CPD.valueOf();
+	private static minimumTokens: number;
+
+	private logger: Logger;
+	private config: Config;
+	private initialized: boolean;
+	private cpdCatalog: Catalog;
+	
+
+
+	public getName(): string {
+		return CpdEngine.ENGINE_NAME;
+	}
+
+	public getTargetPatterns(): Promise<TargetPattern[]> {
+		return this.config.getTargetPatterns(ENGINE.CPD);
+	}
+
+	public getCatalog(): Promise<Catalog>{
+		return Promise.resolve(this.cpdCatalog);
+	}
+
+	/* eslint-disable-next-line no-unused-vars, @typescript-eslint/no-unused-vars */
+	public shouldEngineRun(ruleGroups: RuleGroup[], rules: Rule[], target: RuleTarget[], engineOptions: Map<string, string>): boolean {
+		return true;
+	}
+	
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	public async run(ruleGroups: RuleGroup[], rules: Rule[], targets: RuleTarget[], engineOptions: Map<string, string>): Promise<RuleResult[]> {
+
+		const languageToPaths = this.sortPaths(targets);
+		const results = [];
+
+		if (languageToPaths.size === 0) {
+			this.logger.trace(`No matching cpd target files found. Nothing to execute.`);
+			return [];
+		}
+		
+		// cpd only allows one language to be speificed at a time, so we must run it for each
+		for (const language of Array.from(languageToPaths.keys())) {
+			try {
+				const stdout = await this.runLanguage(language, languageToPaths.get(language));
+				const languageResults = this.processStdOut(stdout)
+				results.push(...languageResults);
+				this.logger.trace(`Found ${languageResults.length} for CPD (${language})`);
+
+			} catch (e) {
+				this.logger.trace(`Cpd (${language}) evaluation failed: ` + (e.message || e));
+				throw new SfdxError(e.message || e);
+			}
+		}   
+		return results;
+	}
+
+	private sortPaths(targets: RuleTarget[]): Map<string, string[]> {
+		const languageToPaths = new Map();
+
+		for (const target of targets) {
+			for (const path of target.paths) {
+				const i = path.lastIndexOf(".");
+				if (i === -1) {
+					//this.logger.warn(`Target: '${path}' was not processed by CPD, no file extension found.`)
+					
+					uxEvents.emit('warning-always', `Target: '${path}' was not processed by CPD, no file extension found.`);
+					continue;
+				}
+				const ext = path.substr(i);
+				if (FILE_EXTS_TO_LANGUAGE.has(ext)) {
+					const language = FILE_EXTS_TO_LANGUAGE.get(ext);
+					const currentPaths = [];
+					if (languageToPaths.has(language)) {
+						currentPaths.push(...languageToPaths.get(language));	
+					} 
+					languageToPaths.set(language, [path, ...currentPaths]);
+				} else {
+					//this.logger.warn(`Cannot run CPD on ${path}, file extension '${ext}' not supported.`)
+					uxEvents.emit('warning-always', `Target: '${path}' was not processed by CPD, file extension '${ext}' not supported.`);
+				}
```

Comment:
In the case that a user adds a target pattern that doesn't match a language we have, which kind of warning do we want: Line 93 or 94?

Reply:
`uxEvents.emit` would make it to UI, whereas `logger.warn` would go to the log file that the user may or may not check. I think `uxEvents` would be more suitable.

Reply:
Makes sense, thanks!